apps/examples/rtc_alarm/rtc_alarm_main.c

```diff
@@ -0,0 +1,353 @@
+
+static void alarm_handler(int signo, FAR siginfo_t *info, FAR void *ucontext)
+{
+	int almndx = info->si_value.sival_int;
+	if (almndx >= 0 && almndx < CONFIG_RTC_NALARMS) {
+		g_alarm_received[almndx] = true;
+	}
+}
+
+/****************************************************************************
+ * Name: alarm_daemon
+ ****************************************************************************/
+
+static int alarm_daemon(int argc, FAR char *argv[])
+{
+	struct sigaction act;
+	sigset_t set;
+	int ret;
+	int i;
+
+	/* Indicate that we are running */
+
+	g_alarm_daemon_started = true;
+	printf("alarm_daemon: Running\n");
+
+	/* Make sure that the alarm signal is unmasked */
+
+	sigemptyset(&set);
+	sigaddset(&set, CONFIG_EXAMPLES_RTC_ALARM_SIGNO);
+
+	ret = sigprocmask(SIG_UNBLOCK, &set, NULL);
+	if (ret != OK) {
+		int errcode = errno;
+		fprintf(stderr, "ERROR: sigprocmask failed: %d\n", errcode);
+		goto errout;
+	}
+
+	/* Register alarm signal handler */
+	act.sa_sigaction = alarm_handler;
+	act.sa_flags     = SA_SIGINFO;
+
+	sigfillset(&act.sa_mask);
+	sigdelset(&act.sa_mask, CONFIG_EXAMPLES_RTC_ALARM_SIGNO);
+
+	ret = sigaction(CONFIG_EXAMPLES_RTC_ALARM_SIGNO, &act, NULL);
+	if (ret < 0) {
+		int errcode = errno;
+		fprintf(stderr, "ERROR: sigaction failed: %d\n", errcode);
+		goto errout;
+	}
+
+	/* Now loop forever, waiting for alarm signals */
+
+	for (; ; ) {
+		/* Check if any alarms fired.
+		*
+		* NOTE that there are race conditions here... if we missing an alarm,
+		* we will just report it a half second late.
+		*/
+		for (i = 0; i < CONFIG_RTC_NALARMS; i++) {
+			if (g_alarm_received[i]) {
+				printf("alarm_daemon: alarm %d received\n", i);
+				g_alarm_received[i] = false;
+			}
+		}
+
+		/* Now wait a little while and poll again.  If a signal is received
+		* this should cause us to awken earlier.
+		*/
+		usleep(500 * 1000L);
+	}
+
+errout:
+	g_alarm_daemon_started = false;
+
+	printf("alarm_daemon: Terminating\n");
+	return EXIT_FAILURE;
+}
+
+/****************************************************************************
+ * Name: start_daemon
+ ****************************************************************************/
+
+static int start_daemon(void)
+{
+	if (!g_alarm_daemon_started) {
+		g_alarm_daemon_pid = task_create("alarm_daemon", CONFIG_EXAMPLES_RTC_ALARM_PRIORITY, CONFIG_EXAMPLES_RTC_ALARM_STACKSIZE, alarm_daemon, NULL);
+		if (g_alarm_daemon_pid < 0) {
+			int errcode = errno;
+			fprintf(stderr, "ERROR: Failed to start alarm_daemon: %d\n", errcode);
+			return -errcode;
+		}
+
+		printf("alarm_daemon started\n");
+		usleep(500 * 1000L);
+	}
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: show_usage
+ ****************************************************************************/
+
+static void show_usage(FAR const char *progname)
+{
+	fprintf(stderr, "USAGE:\n");
+	fprintf(stderr, "\t%s [-a <alarmid>] [-r <alarmid>] [-cr] [<seconds>]\n", progname);
```

Comment:
Let us have this as below.

rtc -alarm [-a <hh:mm> | -r <sec> | -c]

alarm id is not required to be input. Let us use a default id.
If -a is selected, then user must enter absolute time value. Verify that the time entered by user is more than current time from rtc. Then set alarm. 
If -r is selected, user must enter seconds. 
If -c is entered, we will cancel the default alarm id. If default alarm id was never set in the first place, then the ioclt will return error. Then print msg that "no alarm was set"

Reply:
If default alarm id was never set in the first place, rtc ioctl driver is not throwing any error or msg i.e. "no alarm was set". Instead it is showing successful cancellation alarm message.

Reply:
updated usage: alarm [-r seconds] [-a minutes seconds] [-c]
As this is a sample app, so we have kept minutes and seconds for setting rtc alarm with absolute time.

Reply:
How can this be possible `[-a minutes seconds] `
Absolute time must always be in hh:mm:ss format.